packages/flutter_tools/lib/src/macos/xcode.dart

```diff
@@ -185,3 +190,274 @@ class Xcode {
     );
   }
 }
+
+/// A utility class for interacting with Xcode xcdevice command line tools.
+class XCDevice {
+  XCDevice({
+    @required ProcessManager processManager,
+    @required Logger logger,
+    @required Xcode xcode,
+  }) : _processUtils = ProcessUtils(logger: logger, processManager: processManager),
+       _logger = logger,
+       _xcode = xcode;
+
+  final ProcessUtils _processUtils;
+  final Logger _logger;
+  final Xcode _xcode;
+
+  bool get isInstalled => _xcode.isInstalledAndMeetsVersionCheck && xcdevicePath != null;
+
+  String _xcdevicePath;
+  String get xcdevicePath {
+    if (_xcdevicePath == null) {
+      try {
+        _xcdevicePath = _processUtils.runSync(
+          <String>[
+            'xcrun',
+            '--find',
+            'xcdevice'
+          ],
+        ).stdout.trim();
+      } on ProcessException {
+        // Ignored, return null below.
+      } on ArgumentError {
+        // Ignored, return null below.
+      }
+    }
+    return _xcdevicePath;
+  }
+
+  Future<List<dynamic>> _getAllDevices() async {
+    if (!isInstalled) {
+      return null;
+    }
+    try {
+      // USB-tethered devices should always be found.
+      // Finding networked devices takes longer than the default timeout. Try a 5-second timeout.
+      final RunResult result = await _processUtils.run(
+        <String>[
+          xcdevicePath,
+          'list',
+          '--timeout',
+          '5',
+        ],
+      );
+      if (result.exitCode == 0) {
+        return json.decode(result.stdout) as List<dynamic>;
+      }
+      _logger.printTrace('xcdevice returned an error:\n${result.stderr}');
+    } on ProcessException {
+      _logger.printTrace('Failed to invoke xcdevice');
+    }
+
+    return null;
+  }
+
+  /// List of devices USB tethered or paired over the network.
+  /// See Xcode > Window > Devices and Simulators > Connect via network
+  Stream<IOSDevice> getAvailableIOSDevices() async* {
```

Comment:
Why is this a `Stream`?

Reply:
`async*` / `yield` requires this to be a stream.  I can change it to `async` and to just return a `List`, if we prefer that pattern...

Reply:
This could return a `Stream` if we're actually getting results bit-by-bit from `xcdevice`, or if this is implementing an interface that requires it to be a `Stream`, but otherwise we should probably prefer `Future<List>`, or `Future<Iterable>`.

(`async*` `Stream`s also can have some surprising behavior. IIRC they don't interact well with `FakeAsync` for tests.)